src/lib/util/util.js

```diff
@@ -98,6 +99,182 @@ class Util {
+	 */
+	static getFunctionType(fn) {
+		return `(${fn.length}-arity)`;
+	}
+
+	/**
+	 * Get the type of array's elements
+	 * @param {Array} array The array whose element type is to be returned
+	 * @param {number} depth The depth to get the type, recursively
+	 * @param {number} [i=0] Just an iteration count to prevent infinite loops
+	 * @returns {string}
+	 */
+	static getArrayType(array, depth, i = 0) {
+		if (typeof depth !== 'number') throw new TypeError('`depth` is missing, or not a number');
+		return `<${this._getObjType(array, depth, i)}>`;
+	}
+
+	/**
+	 * Get the type of obj's elements
+	 * @param {Object} obj The object whose element type is to be returned
+	 * @param {number} depth The depth to get the type, recursively
+	 * @param {number} [i=0] Just an iteration count to prevent infinite loops
+	 * @returns {string}
+	 */
+	static getObjectType(obj, depth, i = 0) {
+		if (typeof depth !== 'number') throw new TypeError('`depth` is missing, or not a number');
+		const type = this._getObjType(Object.values(obj), depth, i);
+		return type.length > 0 ? `<${this.getComplexType('').type}, ${type}>` : '<>';
+	}
+
+	/**
+	 * Get the type of map's values
+	 * @param {Map} map The map whose value type is to be returned
+	 * @param {number} depth The depth to get the type, recursively
+	 * @param {number} [i=0] Just an iteration count to prevent infinite loops
+	 * @returns {string}
+	 */
+	static getMapType(map, depth, i = 0) {
+		if (typeof depth !== 'number') throw new TypeError('`depth` is missing, or not a number');
+		const keyType = this._getObjType(Array.from(map.keys()), depth, i);
+		const valueType = this._getObjType(Array.from(map.values()), depth, i);
+		return valueType.length > 0 ? `<${keyType}, ${valueType}>` : '<>';
+	}
+
+	/**
+	 * Get the type of set's values
+	 * @param {Set} set The set whose value type is to be returned
+	 * @param {number} depth The depth to get the type, recursively
+	 * @param {number} [i=0] Just an iteration count to prevent infinite loops
+	 * @returns {string}
+	 */
+	static getSetType(set, depth, i = 0) {
+		if (typeof depth !== 'number') throw new TypeError('`depth` is missing, or not a number');
+		return `<${this._getObjType(Array.from(set.values()), depth, i)}>`;
+	}
+
+	/**
+	 * Get the type of values's elements
+	 * @private
+	 * @param {Array} values The array whose element type is to be returned
+	 * @param {number} depth The depth to get the type, recursively
+	 * @param {number} i Just an iteration count to prevent infinite loops
+	 * @returns {string}
+	 */
+	static _getObjType(values, depth, i) {
+		if (!Array.isArray(values)) throw new TypeError("You're using this function wrong; `values` must be an array");
+		if (typeof depth !== 'number') throw new TypeError('`depth` is missing, or not a number');
+		if (typeof i !== 'number') throw new TypeError('`i` is missing, or not a number');
+		// Collections have useful methods, which work on Sets.
+		const Coll = Collection.prototype;
```

Comment:
yeah, let's not....

Reply:
Oh, am I only using that for Collection#first? Yeah, no point in that. (I was originally using another method, as well.)

This whole function is changing in my rewrite, though.

Reply:
fair enough, but lets avoid this code pattern either way